src/tango/ip/fd_ip.h

```diff
@@ -0,0 +1,295 @@
+     buf_cap      the capacity in bytes of buf
+     arp_len      *arp_len is set to the length in bytes of the generated ARP packet
+     ip_addr      the IPv4 address of the target to be probed
+     src_mac_addr the MAC address of the source (caller)
+
+   returns
+     0 on success
+     1 on failure (buf_cap not large enough) */
+
+int
+fd_ip_arp_gen_arp_probe( uchar *   buf,
+                         ulong     buf_cap,
+                         ulong *   arp_len,
+                         uint      dst_ip_addr,
+                         uint      src_ip_addr,
+                         uchar *   src_mac_addr );
+
+
+/* fetch the routing table from the kernel
+
+   the routing table will be written into the workspace, completely replacing
+   any existing routing entries */
+
+void
+fd_ip_route_fetch( fd_ip_t * ip );
+
+
+/* query the routing table
+
+   the provided IP address is looked up in the routing table
+
+   if an appropriate entry is found, *route is set to point to it
+     and 0 is returned
+
+   otherwise, 1 is returned */
+
+int
+fd_ip_route_query( fd_ip_t *              ip,
+                   fd_ip_route_entry_t ** route,
+                   uint                   ip_addr );
+
+
+/* Do routing for an ip_addr
+
+   Handles unicast, broadcast and multicast
+       broadcast:
+         ff.ff.ff.ff -> mac: ff:ff:ff:ff:ff:ff
+       subnet broadcast:
+         a.ff.ff.ff/8
+         a.b.ff.ff/16
+         a.b.c.ff/24 -> if local, ff:ff:ff:ff:ff:ff
+                        else normal routing
+       multicast:
+         224+a,b,c,d -> 01:00:5e:B:c:d (B = b & 0x7f)
+
+   Queries the routing table
+     If no match return -1 error
+     If match,
+       Determines which ip_addr to use
+         If local, uses dst_ip_addr
+         Else, uses the gateway/next hop ip addr
+    Queries the arp table
+      If no match returns 1 - send probe
+      If match, sets mac and ifindex and returns 0
+
+    returns
+      FD_IP_NO_ROUTE  -1  No route to destination
+      FD_IP_SUCCESS    0  Route (and arp if necessary) found. out_* have been set
+      FD_IP_PROBE_RQD  1  Route, but we need to send an ARP probe to resolve the MAC address
+                            Resolve the supplied out_next_ip_addr by sending a probe packet
+      FD_IP_MULTICAST  2  Multicast
+      FD_IP_BROADCAST  3  Local broadcast */
+int
+fd_ip_route_ip_addr( uchar *   out_dst_mac,
+                     uint *    out_next_ip_addr,
+                     uint *    out_ifindex,
+                     fd_ip_t * ip,
+                     uint      ip_addr );
+
+
+/* Prepares the ARP cache for recieving a new ARP entry
+
+   Should be called prior to sending an ARP request
+
+   The kernel ignores unsolicited ARP responses, so this function
+   allows our ARP requests to update the kernel ARP table
+
+   It only adds an entry in the case one does not already exist
+   The entry added is in state "INCOMPLETE", and gets updated to a
```

Comment:
Kind of a nit but sometimes this should be PROBE rather than INCOMPLETE I think? Not sure if there's a difference in the kernel?

Reply:
The state is "incomplete" in netlink